File: ffi/java-jni/java/src/main/java/no/divvun/fst/Speller.java
package no.divvun.fst;

public class Speller implements AutoCloseable {
    private long handle;

    Speller(long handle) {
        this.handle = handle;
    }

    public boolean isCorrect(String word) {
        if (handle == 0) {
            throw new IllegalStateException("Speller is closed");
        }
        if (word == null) {
            throw new NullPointerException("word cannot be null");
        }
        return isCorrect(handle, word);
    }

    public SuggestionList suggest(String word) {
        if (handle == 0) {
            throw new IllegalStateException("Speller is closed");
        }
        if (word == null) {
            throw new NullPointerException("word cannot be null");
        }
        long suggestionsHandle = suggest(handle, word);
        if (suggestionsHandle == 0) {
            throw new RuntimeException("Failed to get suggestions");
        }
        return new SuggestionList(suggestionsHandle);
    }

    @Override
    public void close() {
        if (handle != 0) {
            free(handle);
            handle = 0;
        }
    }

    private static native boolean isCorrect(long handle, String word);
    private static native long suggest(long handle, String word);
    private static native void free(long handle);
}

File: ffi/swift/Sources/DivvunFST/Speller.swift
import Foundation
import CDivvunFST

public class Speller {
    private var speller: DFST_Speller

    init(speller: DFST_Speller) {
        self.speller = speller
    }

    public func isCorrect(_ word: String) throws -> Bool {
        let exceptionCallback: cffi_exception_callback = { msg, msgLen in
        }

        let wordBytes = Array(word.utf8)
        let result = wordBytes.withUnsafeBytes { wordPtr in
            let wordSlice = rust_slice_t(data: UnsafeMutableRawPointer(mutating: wordPtr.baseAddress), len: UInt(wordBytes.count))
            return DFST_Speller_isCorrect(speller, wordSlice, exceptionCallback)
        }

        return result != 0
    }

    public func suggest(_ word: String) throws -> SuggestionList {
        let exceptionCallback: cffi_exception_callback = { msg, msgLen in
        }

        let wordBytes = Array(word.utf8)
        let suggestions = wordBytes.withUnsafeBytes { wordPtr in
            let wordSlice = rust_slice_t(data: UnsafeMutableRawPointer(mutating: wordPtr.baseAddress), len: UInt(wordBytes.count))
            return DFST_Speller_suggest(speller, wordSlice, exceptionCallback)
        }

        if suggestions.data == nil {
            throw NSError(domain: "DivvunFST", code: 6, userInfo: [NSLocalizedDescriptionKey: "Failed to get suggestions"])
        }

        return SuggestionList(handle: suggestions)
    }
}
